### PR TITLE
Refactor arming checks to events interface

### PR DESCRIPTION
**Pull request by bkueng**
This finally brings the arming checks to the events interface from https://github.com/PX4/PX4-Autopilot/pull/16293.

I changed a few things for added flexibility. Arming checks are reported per mode group, and these groups we can now freely define. I'm using 1 mode = 1 group now, so all modes can report different checks. We can also define the components entirely in PX4 now.

Minimual UI in QGC:

https://user-images.githubusercontent.com/281593/183851356-578353a1-4f81-4467-b093-0bc17ab6c691.mp4

### Check Changes
- dropped the v2 vtol airframe check
- dropped the accel magnitude check (was based on a single sample, we can add a more robust check)
- airspeed check on arming attempt - it's now checking always (if enabled) but only as a warning
- arm authorization is called more often
- checks run & report while armed too

### Compatibility & console output
mavlink_log is still published for backwards compatibility. This is based on 2 passes: whenever there's a change in a check, the checks are run again with mavlink log reporting enabled.
@dagar this still prints to the console, and when removing those we can enable `CONSOLE_PRINT_ARMING_CHECK_EVENT` if you prefer. It will output the reports whenever they change in the form of:
```
Sending Arming/Health Report (num events: 4, can_arm: 0xffffbff7, can_run: 0xffffbff7)
   Event 0x0105c2d9: No mission available
   Event 0x012f1dc7: No offboard signal
```

Later on I'd also like to have a MAVSDK CLI tool to print the report. @julianoes any interest in that?

### Mode requirements
This also adds the concept of mode requirements. It abstracts mode-specifics. Initially I did not want to add that (yet), but it makes things a bit simpler later on.
Information flow:
```
   mode requirements     --->  arming checks & reporting     ---> set can_run bits    ---> mode switching check
                         ---> failsafe state machine
```

### Performance
on px4-fmu-v5:
- RAM: an increase of 3KB, and the first commit (https://github.com/PX4/PX4-Autopilot/commit/aba72e35eef541f3584d7dabad2f040b1649da7a) reduces by 3KB, so the PR is a net zero.
- FLASH: it's not cheap, it adds ~11KB (5KB are the mavlink_log messages)
- CPU: the checks run much faster, 0.75ms vs 6.2ms on main

### SYS_STATUS
The check also replace the SYS_STATUS setting, and there's translation to publish the existing mavlink message. There's a few changes, in particular `SUBSYSTEM_TYPE_SATCOM` got removed. @julianoes is there anything MAVSDK relies on? Preflight check (for the current mode) is still reported through that.

### Testing
- backwards compatibility test
- tested on a vehicle with gps & sitl, with different configurations
- tested on an unconfigured system

Requires https://github.com/mavlink/libevents/pull/2

### CODE REVIEW

**Comment by julianoes** [comment]
> Later on I'd also like to have a MAVSDK CLI tool to print the report. @julianoes any interest in that?

Yes.

> @julianoes is there anything MAVSDK relies on? 

MAVSDK currently looks at:
- `MAV_SYS_STATUS_SENSOR_RC_RECEIVER`
- `MAV_SYS_STATUS_SENSOR_3D_GYRO`
- `MAV_SYS_STATUS_SENSOR_3D_ACCEL`
- `MAV_SYS_STATUS_SENSOR_3D_MAG`
- `MAV_SYS_STATUS_SENSOR_GPS`
- `MAV_SYS_STATUS_SENSOR_OPTICAL_FLOW`
- `MAV_SYS_STATUS_SENSOR_VISION_POSITION`

https://github.com/mavlink/MAVSDK/blob/425577aa5629d591a854a65fd1f2dd6d88b526d9/src/mavsdk/plugins/telemetry/telemetry_impl.cpp#L1086-L1121

And it determines armable based on `MAV_SYS_STATUS_PREARM_CHECK`

https://github.com/mavlink/MAVSDK/blob/425577aa5629d591a854a65fd1f2dd6d88b526d9/src/mavsdk/plugins/telemetry/telemetry_impl.cpp#L1142-L1145

**Comment by bkueng** [comment]
> MAVSDK currently looks at:

Thanks, that should be ok then. The sensors unhealthy will be set in a few more cases now, such as driver stops publishing or no valid calibration data. Previously it was only set when the estimator declared it unhealthy.

**Comment by bkueng** [comment]
CI is passing, I found a number of things to reduce flash usage.

File: src/modules/commander/HealthAndArmingChecks/HealthAndArmingChecks.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```
Comment (bresch) [inline src/modules/commander/HealthAndArmingChecks/HealthAndArmingChecks.cpp:3]:
```suggestion
 *   Copyright (c) 2022 PX4 Development Team. All rights reserved.
```

Comment (bkueng) [reply src/modules/commander/HealthAndArmingChecks/HealthAndArmingChecks.cpp:3]: Believe it or not, this is when it was actually created

**Comment by bkueng** [comment]
@dagar good to merge?

**Review by dagar — APPROVED** [review]

**Review by dagar — APPROVED** [review]

**Comment by hamishwillee** [comment]
@bkueng Very cool. Given how important this is, we should perhaps document this in places we talk about arming and modes.

My understanding is:
1. **Anything** that prevents arming will result in an event in appropriate version of QGC - correct? I.e. there is never going to be a case where the vehicle doesn't arm and you can't find out from the arming UI what you need to do?
   - What causes UI update? Say the the safety button is pressed, or some param interlock is preventing arming - is it that you're running the arming check on some kind of timer, and that is what makes the UI update, or is it that pressing the button or changing a param is what triggers an arming check?
2.  Previously you could only switch to particular modes when disarmed - is that right? What were they?
3. Now you can switch to any mode in disarmed. If you try to arm though the armed check is run and you'll get mode-specific reasons for failure - is that right?
4. There were some comments above - what does it mean?

   > The check also replace the SYS_STATUS setting, and there's translation to publish the existing mavlink message. There's a few changes, in particular SUBSYSTEM_TYPE_SATCOM got removed. @julianoes is there anything MAVSDK relies on? Preflight check (for the current mode) is still reported through that.
   
   - Does this mean that PX4 is no longer using https://mavlink.io/en/messages/common.html#MAV_SYS_STATUS_SENSOR to indicate the status of sensors? 
   - Or that everything relevant in https://mavlink.io/en/messages/common.html#MAV_SYS_STATUS_SENSOR is now in events? 
   - Or something else?

5. If something else wants to implement support for this, where would they look for information on what might be emitted and how to interpret it?

**Comment by hamishwillee** [comment]
Thanks @bkueng - this is a much needed thing you have done. 

- I am pretty sure the virtual joystick thing is a valid bug - posted in https://github.com/PX4/PX4-Autopilot/issues/21953
- FYI warnings like "Preflight: GPS Vertical Pos Drift too high" tell me I probably need a better GPS fix, but a "new user" would probably benefit from some instruction like "This may improve if you wait for a few seconds. If not, move vehicle into an unobstructed position". I guess I'm asking if there is any thinking around "what's the right level for error messages?"
- I updated QGC docs for this in [Fly View](https://docs.qgroundcontrol.com/master/en/FlyView/FlyView.html#arm) (https://github.com/mavlink/qgc-user-guide/pull/398). Looks like no one is doing QGC docs now days.
- Updated Basic concepts to have a cross link on arming, and also to fix up the old material on actuator/outputs. No need to particularly review.
- I still want to add a section in the main arming/prearming page. Next week.

**Comment by bkueng** [comment]
> FYI warnings like "Preflight: GPS Vertical Pos Drift too high" tell me I probably need a better GPS fix, but a "new user" would probably benefit from some instruction like "This may improve if you wait for a few seconds. If not, move vehicle into an unobstructed position". I guess I'm asking if there is any thinking around "what's the right level for error messages?"

Yes we should definitely take another pass at this. You're welcome to do so. We can also add links to docs.

> I updated QGC docs for this in [Fly View](https://docs.qgroundcontrol.com/master/en/FlyView/FlyView.html#arm) (https://github.com/mavlink/qgc-user-guide/pull/398). Looks like no one is doing QGC docs now days.

Cool, thanks.

**Comment by hamishwillee** [comment]
> Yes we should definitely take another pass at this. You're welcome to do so. We can also add links to docs.

This needs to be done by someone who knows what they are doing in order to understand the errors and actions. I did what can be guessed "back in the day" here: https://docs.px4.io/main/en/flying/pre_flight_checks.html#preflight-sensor-estimator-checks

I would prefer this in the code and having developers getting into the habit of writing it for end users as they create new checks. 
Very happy to subedit though.

**Pull request merged**